### PR TITLE
Support supplemental code points while scanning

### PR DESCRIPTION
**Pull request by caitp**
since the unicode tables deal with supplemental codepoints, it is probably a good idea to actually scan them!

good things to do: for identifiers, use a generated identifier in codegen instead, so that it might work in legacy engines.

### CODE REVIEW

File: src/syntax/Scanner.js [hunk 1]
```diff
@@ -1172,11 +1173,30 @@ function isAtEnd() {
...
   index++;
-  updateCurrentCharCode();
+  if (currentCodePoint >= 0x10000) {
+    index++;
+  }
+  updateCurrentCodePoint();
+}
+
+function updateCurrentCodePoint() {
+  currentCodePoint = currentCharCode = input.charCodeAt(index);
```
Comment (arv) [inline src/syntax/Scanner.js]: is currentCharCode still used?

File: src/syntax/Scanner.js [hunk 2]
```diff
@@ -1172,11 +1173,30 @@ function isAtEnd() {
...
+  nextCharCode = input.charCodeAt(index + 1) | 0;
```
Comment (arv) [inline src/syntax/Scanner.js]: This is really hot code. What is the perf impact?

Comment (arv) [inline src/syntax/Scanner.js]: nextCharCode is only used in this function. Use a local variable.

File: src/syntax/Scanner.js [hunk 3]
```diff
@@ -1172,11 +1173,30 @@ function isAtEnd() {
...
+  if (isLeadSurrogate(currentCodePoint) && isTrailSurrogate(nextCharCode)) {
```
Comment (arv) [inline src/syntax/Scanner.js]: One simple optimization would probably be to to check if the currentCodePoint is ASCII.

```js
function updateCurrentCodePoint() {
  currentCodePoint = input.charCodeAt(index);
  if (currentCodePoint < 128) return;
  ...
}
```

**Comment by mathiasbynens** [comment]
Small stylistic nit: how about using consistently using hexadecimal integer literals to refer to code points?

E.g.:

```js
// instead of…
currentCodePoint === 47
// use:
currentCodePoint === 0x2F
```

This makes it much clearer that this is about U+002F, IMHO.

**Comment by arv** [comment]
@mathiasbynens Maybe unicode geeks like you dream in hex but I still operate better with decimals.

**Comment by mathiasbynens** [comment]
@arv It’s just inconsistent right now. See `0xD800` at <https://github.com/google/traceur-compiler/pull/1824/files#diff-05e5b7795cfe69ccb6421b981250effbR1191> and `0xDC00` at <https://github.com/google/traceur-compiler/pull/1824/files#diff-05e5b7795cfe69ccb6421b981250effbR1195>.

**Comment by caitp** [comment]
I would recognize 0xFFFE easier than 65534, but that may not apply to decimals in the ascii range

**Comment by mathiasbynens** [comment]
@caitp Pfff, double standards! :P

File: src/syntax/Scanner.js [hunk 4]
```diff
@@ -1172,11 +1172,33 @@ function isAtEnd() {
...
+    // If code point is higher than Plane 0, move past trail surrogate
+    index++;
+  }
+  updateCurrentCodePoint();
+}
+
+function updateCurrentCodePoint() {
+  currentCodePoint = input.charCodeAt(index);
+  // Return if code point is in Plane 0
+  if (currentCodePoint < 0xD800 || !isLeadSurrogate(currentCodePoint)) return;
```
Comment (caitp) [inline src/syntax/Scanner.js]: fast case: U+D7FF is a pretty big range of codes, and is the last code point before a trail of lead surrogates --- slightly slower are at the higher end of the BMP, which are detected with the function call

File: src/syntax/Scanner.js [hunk 5]
```diff
@@ -1172,11 +1172,33 @@ function isAtEnd() {
...
+  let nextCharCode = input.charCodeAt(index + 1) | 0;
+  // ASSERT: isTrailSurrogate(nextCharCode)
+  // Should report an error if it's not?
```
Comment (caitp) [inline src/syntax/Scanner.js]: could get rid of `isTrailSurrogate()` if we never care about this, I guess

**Comment by arv** [comment]
@caitp Same Q here. Do you want to continue with this?

**Comment by caitp** [comment]
did we decide it was worth it? it's a performance hit for probably no practical purpose, but you should make the call

**Comment by arv** [comment]
Closing this one too...

**Pull request closed without merging**